### PR TITLE
Port plot_slice to typescript

### LINKED ISSUE #28: [Feature] Support plot_slice to show the parameter relationship.
# Feature Request

Porting `plot_slice()` to TypeScript.
https://github.com/optuna/optuna/blob/v2.4.0/optuna/visualization/_slice.py

### PR DESCRIPTION
**Pull request by 2403hwaseer**
# Description
Porting `optuna.visualization.plot_slice` to typescript. Users can now visualize slice plot in dashboard.

# Context
Resolves #28

### CODE REVIEW

**Comment by 2403hwaseer** [comment]
> TS2322: Type 'undefined' is not assignable to type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }[]'.

> TS2322: Type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }[]' is not assignable to type 'Partial<PlotData>[]'.
>   Type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }' is not assignable to type 'Partial<PlotData>'.
>     Types of property 'type' are incompatible.
>       Type 'string' is not assignable to type '"table" | "image" | "scatter" | "bar" | "barpolar" | "box" | "candlestick" | "carpet" | "choropleth" | "choroplethmapbox" | "cone" | "contour" | "contourcarpet" | "contourgl" | ... 33 more ... | undefined'.
> 

I am getting the above mentioned erros while compiling and I am not able to figure out how to eliminate these errors. Can you please help me out?

File: optuna_dashboard/static/components/GraphSlice.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    else{
+      let i=0
+      let  trace =  {
+        type: "scatter",
+        dimensions :dimensions, 
+        mode :"markers",
+        xaxis : "x"
+      }
+
+       let traces :{type: string, dimensions: { label: string; values: number[]; range: number[]; }[], mode: string, xaxis: string} [] = undefined
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphSlice.tsx:88]: > TS2322: Type 'undefined' is not assignable to type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }[]'.

You can fix the above error like this:

```suggestion
       let traces :{type: string, dimensions: { label: string; values: number[]; range: number[]; }[], mode: string, xaxis: string} [] = []
```

Comment (c-bata) [reply optuna_dashboard/static/components/GraphSlice.tsx:88]: And to resolve the following error, you need to type `traces: Partial<plotly.PlotData>[]`.

> TS2322: Type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }[]' is not assignable to type 'Partial<PlotData>[]'.
>   Type '{ type: string; dimensions: { label: string; values: number[]; range: number[]; }[]; mode: string; xaxis: string; }' is not assignable to type 'Partial<PlotData>'.
>     Types of property 'type' are incompatible.
>       Type 'string' is not assignable to type '"table" | "image" | "scatter" | "bar" | "barpolar" | "box" | "candlestick" | "carpet" | "choropleth" | "choroplethmapbox" | "cone" | "contour" | "contourcarpet" | "contourgl" | ... 33 more ... | undefined'.

**Comment by 2403hwaseer** [comment]
Hey @c-bata , I tried but I am unable to figure out why is the `GraphSlice.tsx` component displaying an empty graph. Can you suggest me any changes?

**Comment by c-bata** [comment]
I'm sorry. I may not be able to support anymore :bow:

**Comment by 2403hwaseer** [comment]
@c-bata okay, I understand. In order to complete the pending work, can you please connect me with someone who can guide me on this?

**Comment by c-bata** [comment]
optuna-dashboard has been developed by me and there is no other contributors. So seemingly it is also difficult to connect someone with you. I'm sorry.

**Comment by 2403hwaseer** [comment]
Hey @c-bata , thanks for the link. I did study about plotly.js but I think I might have missed something. I'll have a look at it and I'll try to resolve this by myself.

**Comment by 2403hwaseer** [comment]
Hey @c-bata , I made some changes to my code and the graph is now visible. However, I was facing a lot of difficulty in updating the `xaxis` property of `Partial<plotly.Layout>`.

For example, if there are 3 parameters to be plotted, I can update the `xaxis` , `xaxis2` and `xaxis3` property manually. But what if there are 100 parameters?

I tried writing the following code:
```
let i=1
paramNames.foreach((paramName) => {
      if(i!=1){
        let axisname = `xaxis${i}`
        updatelayout[axisname] = {
          title: paramName,
          zerolinecolor: "#f2f5fa",
          zerolinewidth: 1.5,
          linecolor: "#f2f5fa",
          inewidth: 5,
          gridcolor: "#f2f5fa",
          gridwidth:1,

        }
      }
      i++
 })
```
However, it gives the following error 

> Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Partial<Layout>'.
>   No index signature with a parameter of type 'string' was found on type 'Partial<Layout>'.

```

```

**Comment by c-bata** [comment]
Hi,

Thank you for the updates. I confirmed that the graph is now visible like this:

Screenshot 2021-02-22 20 10 23

> For example, if there are 3 parameters to be plotted, I can update the xaxis , xaxis2 and xaxis3 property manually. But what if there are 100 parameters?

How about putting the select box or radio button to choose the parameter like there is a radio button on the left of history plot?

**Comment by 2403hwaseer** [comment]
Hi @c-bata ,Yes, I think it would be a good idea to add a select box so that the graph does not look compact. Once I have implemented this, I'll request for a review again.

File: optuna_dashboard/static/components/GraphSlice.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import * as plotly from "plotly.js-dist"
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphSlice.tsx:1]: I replaced `plotly.js-dist` with `plotly.js-basic-dist` at https://github.com/optuna/optuna-dashboard/pull/40 to reduce the file size of bundle.js. Could you merge `master` branch and replace here like below?

```suggestion
import * as plotly from "plotly.js-basic-dist"
```

**Comment by 2403hwaseer** [comment]
Hi @c-bata , I have added the Select Button in `GraphSlice.tsx` . Can you please take a look?

**Comment by c-bata** [comment]
https://user-images.githubusercontent.com/5564044/108921473-c0a39b00-7679-11eb-94c9-9bb73269e32d.mp4

Thank you for the updates. I have some suggestions.

* I found a bug that the graph does not support categorical parameters. Could you fix this?
    * Optuna's `plot_slice()` function supports categorical parameters like this: ScreenShot 2021-02-24 8 29 16
* It looks that the column width of select box is a half of the container width. I think a quarter of the container width is enough for select box area like history plot. What do you think?
    * ScreenShot 2021-02-24 8 34 52
* How about setting the initial value of select box? Otherwise users cannot imagine whether the graph of `plot_slice()` will be shown.

**Comment by 2403hwaseer** [comment]
Hey @c-bata , thanks for the review. Here are some comments of mine:
1.  I did not test the code using categorical parameters so thank you for reporting this. I will add the support for this and let you know.
2. So, with respect to the width of select box, it's actually 1/4th of the container  but due to the alignment of the graph, it visually feels like it's half of the container width. I'll fix the positioning of these elements.
3. Oh yes, I actually set the initial param value to x and didn't realise that it won't be necessarily true that initial value defaults to x every time. I will fix this too!

**Comment by 2403hwaseer** [comment]
Hi @c-bata , I have added support for categorical parameters. The x-axis labels are not visible and I am trying to figure out why. Will let you know once done.

**Comment by 2403hwaseer** [comment]
Hi @c-bata , I think this is ready now. Can you please take a look?

**Comment by c-bata** [comment]
Thank you for the update! I confirmed that the bug of categorical parameter is fixed.
In my environment, the graph does not update even though select another ones. Could you fix this?

https://user-images.githubusercontent.com/5564044/109093794-4ac43000-775c-11eb-9120-f3008c1b3ef5.mp4

**Comment by 2403hwaseer** [comment]
Hi @c-bata , I have made the necessary changes. I think the graph should now display the plots of all parameters.

**Comment by 2403hwaseer** [comment]
Sorry. It got closed by mistake.

**Review by c-bata — APPROVED** [review]
LGTM! Thank you for your contribution 👍

**Pull request merged**